### PR TITLE
Roadmapping for 2023

### PR DESCRIPTION
**Pull request by rvagg**
This comes out of a discussion happening in Notion, which is not public, so I want to surface this for broader input and also in a format that we may be all more comfortable having discussion in.

This is my attempt at combining my thoughts on a roadmap with input that I've seen so far from @RangerMauve @willscott and @masih.

* @RangerMauve has an excellent visioning post up at https://blog.mauve.moe/posts/ipld-vision
* @masih has been developing a very mature and detailed set of thoughts at https://hackmd.io/KgnTcWefToqrXHrYU4qqEA

This doc is incomplete and a little bit rushed, so don't expect too much of it but I'd love to hear what folks have to say about my synthesis.

TODO includes:

* Prioritisation
* More details in some areas that lack it
* Milestones that are practical within various resourcing constraints - what could we get done with various people configurations, including: "Status quo", "Lightly staffed", and "Team of 2-3 people".

### CODE REVIEW

**Comment by rvagg** [comment]
I've only been in Lisbon for a little over 12 hours and I've already had multiple discussions about IPLD & Rust as they relate to roadmap items. It's been looming larger in my mind as a major theme for 2023; FVM and Iroh are both important users that need solid support. Because it's such a greenfield it may also be a chance to revisit some basic API ideas too.

Anyway, this is the current state of a quick roadmap I want to present internally that may feed in to resourcing decisions. We've got a few days before this is presented and I'm trying to arrange a roadmapping session on Wednesday morning Lisbon time (which should have a Zoom so put your hand up if you're reading this and would like to join in person or via Zoom), and this slide can be updated afterward. It's not an authoritative list because the resourcing questions are the big ones that need to be addressed first, but I see this as a chance to pain a possible vision for what we _could_ be doing if we had more energy and people-time dedicated to IPLD in the coming year.

Screenshot 2022-10-23 at 11 13 37 am

Feedback welcome!

File: ROADMAP.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+## Focus Areas
+
+### Documentation
+
+1. Improved onboarding documentation and resources.
+2. Improved UX for consolidated docs/specs site (ipld.io)
+
+### Distributed Data Structures
+
+Specifications and documentation related to data structures to represent common application data over content addressed data. Including: filesystem data, ordered and unordered maps, sparse and dense arrays, encrypted and/or signed data.
```
Comment (masih) [inline ROADMAP.md:25]: How is this different from Documentation above? Maybe worth renaming the "Documentation" section title to something like Onboarding and Usability?

File: ROADMAP.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+   4. Data structure fixtures: are currently very minimal and mainly provided as an example and not consumed from ipld/ipld by implementations.
+   5. CID & core multiformats fixtures: need collecting and documenting in an appropriate place (ipld/ipld may not be the right location for these)
+
+### CAR Consolidation
+
+There has been an explosion of activity around CAR as a fundamental data format. This activity is not well reflected in core documentation or specifications. Investment is needed to document, consolidate and shepherd CAR-related activity in order to ensure maximum compatibility throughout the ecosystem and provide a pathway to solutions that various parties are seeking through the use of CAR, CAR extensions and CAR-related tooling.
+
+This work may also involve a CARv3, or (more likely) extensions of CARv2 to support user-needs.
+
+### Implementation Specific: Go
```
Comment (masih) [inline ROADMAP.md:54]: Worth mentioning consolidation of HAMT implementations?

**Comment by ec1oud** [comment]
UnixFS v1 is still holding things back.  We need a way to attach arbitrary IPLD metadata to files, in a way that most parts of the ecosystem can understand, without having to write special ADLs and special application-specific wasm plugins and so on.  It doesn't seem like it should be so hard, but I haven't been able to find any mainstream file-adding APIs that can take some IPLD along with the files and directories.  On Linux we have extended attributes, and on IPFS we need at least that... but hopefully much more.

File: ROADMAP.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+Next steps include:
+
+1. Expanding codec test fixturing to include *negative* tests which can strictly describe the boundaries of correctness for codec implementations; i.e. data forms at the boundaries which are expected to fail encoding or decoding using specific codecs.
+2. Further consolidation of test fixtures into the ipld/ipld repository and consumption of this data as a source of truth. This likely includes investment in [testmark](https://github.com/warpfork/go-testmark) style fixtures.
+   1. Schema fixtures: not consistently consumed across JavaScript and Go from ipld/ipld
+   2. Codec fixtures: source of truth is ipld/codec-fixtures; current representation as testmark in ipld/ipld is not ideal.
+   3. CAR fixtures: live in js-car and go-car, with some overlap.
+   4. Data structure fixtures: are currently very minimal and mainly provided as an example and not consumed from ipld/ipld by implementations.
+   5. CID & core multiformats fixtures: need collecting and documenting in an appropriate place (ipld/ipld may not be the right location for these)
+
```
Comment (RangerMauve) [inline ROADMAP.md:47]:
```suggestion
  6. ADL fixtures: Even though we have specs for ADLs like HAMT, they are not consistently implemented in some cases and generally have their own testing suites. We should have a more rigorous standard for ADL spec + implementations, and at least get HAMT/ProllyTrees consistent across major implementations

```

File: ROADMAP.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+### Implementation Specific: JavaScript
+
+* Possible Node interface(s) to allow better/full support of core functionality available in Go
+* Lenses and layering
+* APIs to pull them together
+
+### Implementation Specific: Rust
+
+The current Rust tooling is weak and incomplete. Investment is required to complete basic components (e.g. codecs) and explore possibilities for improved interfaces, possibly similar to that which is available in Go today.
```
Comment (RangerMauve) [inline ROADMAP.md:87]:
```suggestion
The current Rust tooling is weak and incomplete. Investment is required to complete basic components (e.g. codecs) and explore possibilities for improved interfaces, possibly similar to that which is available in Go today. This is especially important with regards to the growing number of projects in the PL ecosystem that are using Rust. At the moment, even if a project wanted to adopt IPLD features, there's no easy path for them to take.
```

**Comment by willscott** [comment]
These are a reasonable set of technical focus areas. I would suggest that growing team capacity should be called out here explicitly, because without spending time and resources on that it's unlikely that intended capacity gets made on any of the specific work items proposed here.
